joint_state_controller/src/joint_state_controller.cpp

```diff
@@ -84,93 +102,101 @@ JointStateController::on_deactivate(const rclcpp_lifecycle::State & /*previous_s
 {
-  if (auto rh_ptr = robot_hardware_.lock()) {
-    joint_names_ = rh_ptr->get_registered_joint_names();
+  bool found_key = false;
+  for (const auto & key_item : map) {
+    const auto & key = key_item.first;
+    if (std::find(keys.cbegin(), keys.cend(), key) != keys.cend()) {
+      found_key = true;
+      break;
+    }
+  }
+  return found_key;
+}
 
-    if (joint_names_.empty()) {
-      return false;
+bool JointStateController::init_joint_data()
+{
+  for (const auto & state_interface : state_interfaces_) {
+    // initialize map if name is new
+    if (name_if_value_mapping_.count(state_interface.get_name()) == 0) {
+      name_if_value_mapping_[state_interface.get_name()] = {};
     }
+    // add interface name
+    name_if_value_mapping_[state_interface.get_name()][state_interface.get_interface_name()] =
+      kUninitializedValue;
+  }
 
-    // reset any previous joint data
-    joint_handles_.clear();
-
-    for (const auto & joint_name : joint_names_) {
-      // always check for these interfaces, even if they are not supported by the joint
-      std::vector<std::string> interface_names = {"position", "velocity", "effort"};
-      const std::vector<std::string> joint_interface_names =
-        rh_ptr->get_registered_joint_interface_names(joint_name);
-
-      // add the rest of the interfaces
-      // ideally this could be done with std::sort + std::unique, but we rather have
-      // position, velocity and effort in known indexes
-      std::copy_if(
-        joint_interface_names.begin(), joint_interface_names.end(),
-        std::back_inserter(interface_names), [&](const auto & value) {
-          return std::find(
-            interface_names.begin(),
-            interface_names.end(), value) == interface_names.end();
-        });
-
-      std::vector<std::shared_ptr<hardware_interface::JointHandle>> joint_interface_handles;
-      for (const auto & interface_name : interface_names) {
-        /// @todo is there a better way of skipping command interfaces?
-        if (interface_name.find("command") != std::string::npos) {
-          continue;
-        }
-
-        auto joint_handle = std::make_shared<hardware_interface::JointHandle>(
-          joint_name,
-          interface_name);
-        if (rh_ptr->get_joint_handle(*joint_handle) == hardware_interface::return_type::OK) {
-          joint_interface_handles.emplace_back(joint_handle);
-        } else {
-          // leave a nullptr for not supported interfaces
-          joint_interface_handles.emplace_back(nullptr);
-        }
-      }
-      joint_handles_.emplace_back(joint_interface_handles);
+  // filter state interfaces that have at least one of the joint_states fields,
+  // the rest will be ignored for this message
+  for (const auto name_ifv : name_if_value_mapping_) {
+    const auto & interfaces_and_values = name_ifv.second;
+    if (has_any_key(interfaces_and_values, {kPositionName, kVelocityName, kEffortName})) {
```

Comment:
```suggestion
    if (has_any_key(interfaces_and_values, {hardware_interface::HW_IF_POSITION,
                                                                     hardware_interface::HW_IF_VELOCITY,
                                                                     hardware_interface::HW_IF_EFFORT})) {
```

Maybe not the best formatting possible...

Reply:
when working around linters and code formatters, there's only so much you can discuss about code style :D 

removing the hardware_interface namespace by a couple of `using`s at the top of the file kept this a one-liner

Reply:
Yeah, I mean my proposal was not the best formating possible.